src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexNodes.cs

```diff
@@ -0,0 +1,1643 @@
+            }
+
+            throw new InvalidOperationException();
+        }
+
+        public override void Accept(IRegexNodeVisitor visitor)
+            => visitor.Visit(this);
+    }
+
+    /// <summary>
+    /// ```expr+```
+    /// </summary>
+    internal sealed class RegexOneOrMoreQuantifierNode : RegexQuantifierNode
+    {
+        public RegexOneOrMoreQuantifierNode(
+            RegexExpressionNode expression, RegexToken plusToken)
+            : base(RegexKind.OneOrMoreQuantifier)
+        {
+            Debug.Assert(expression != null);
+            Debug.Assert(plusToken.Kind == RegexKind.PlusToken);
+            Expression = expression;
+            PlusToken = plusToken;
+        }
+
+        public RegexExpressionNode Expression { get; }
+        public RegexToken PlusToken { get; }
+
+        internal override int ChildCount => 2;
+
+        internal override RegexNodeOrToken ChildAt(int index)
+        {
+            switch (index)
+            {
+            case 0: return this.Expression;
+            case 1: return this.PlusToken;
+            }
+
+            throw new InvalidOperationException();
+        }
+
+        public override void Accept(IRegexNodeVisitor visitor)
+            => visitor.Visit(this);
+    }
+
+    /// <summary>
+    /// ```expr?```
+    /// </summary>
+    internal sealed class RegexZeroOrOneQuantifierNode : RegexQuantifierNode
+    {
+        public RegexZeroOrOneQuantifierNode(
+            RegexExpressionNode expression, RegexToken questionToken)
+            : base(RegexKind.ZeroOrOneQuantifier)
+        {
+            Debug.Assert(expression != null);
+            Debug.Assert(questionToken.Kind == RegexKind.QuestionToken);
+            Expression = expression;
+            QuestionToken = questionToken;
+        }
+
+        public RegexExpressionNode Expression { get; }
+        public RegexToken QuestionToken { get; }
+
+        internal override int ChildCount => 2;
+
+        internal override RegexNodeOrToken ChildAt(int index)
+        {
+            switch (index)
+            {
+            case 0: return this.Expression;
+            case 1: return this.QuestionToken;
+            }
+
+            throw new InvalidOperationException();
+        }
+
+        public override void Accept(IRegexNodeVisitor visitor)
+            => visitor.Visit(this);
+    }
+
+    /// <summary>
+    /// Quantifiers can be optionally followed by a ? to make them lazy.  i.e. ```a*?``` or ```a+?```.
+    /// You can even have ```a??```  (zero or one 'a', lazy).  However, only one lazy modifier is allowed
+    /// ```a*??``` or ```a???``` is not allowed.
+    /// </summary>
+    internal sealed class RegexLazyQuantifierNode : RegexExpressionNode
+    {
+        public RegexLazyQuantifierNode(
+            RegexQuantifierNode quantifier, RegexToken questionToken)
+            : base(RegexKind.LazyQuantifier)
+        {
+            Debug.Assert(quantifier != null);
```

Comment:
Consider asserting that `quantifier.Kind != RegexKind.LazyQuantifier` as described in comment.

Reply:
#resolved